rfc/helm.md

```diff
@@ -0,0 +1,131 @@
+                                                                                                                                                                                                                                             
+Helm charts can't rely on go code besides the functions provided by the go template library, the Sprig library, and a few extra functions provided by Helm itself. This implies a few compromises:
+                                                         
+- We can't validate the install options provided in `values.yaml`. Instead, a new set of `linkerd check` checks could help catching invalid options, post-install.
+- We should provide a comprehensive `values.yaml` file that would contain the most common settings, but heavily annotated to instruct users about alternate settings for advanced scenarios.                                                              
+- Helm's crypto functions only allow us to use RSA certs/keys. We can move the cert/keys from the webhook configs (`proxy-injector` and `sp-validator`) to use RSA. As for the trust root for identity, we decided the user should provide their own in `values.yaml`. The docs should have instructions on how to generate that cert/key.
+   
+These compromises entail a less straightforward experience than what `linkerd install` provides, so the Helm installation alternative should be considered an "advanced" feature.                                                             
+                                                                                                                                                                                                                                             
+New alternative install workflow through Helm
+-----------------------------------------------
+```               
+helm install incubator/linkerd2
+```                                                                                                                                                                                                                                          
+That would install linkerd using the most common settings. The `NOTES.txt` file (rendered and shown when this command completes) could provide the follow instructions/warnings:
+- Warn that the identity trust root should have been provided, and show instructions on how to generate it.
+- Instructions on how to, optionally, install the linkerd CLI
+- Instructions on running `linkerd check` to verify everything is ok                                                                                                                                                                         
+- Instructions on how to change the most basic settings
+- Instructions on how to get ahold of `values.yaml` containing all the possible settings?
+
+### New Chart Layout
+This will replace the current single chart under the `charts` directory.
+```
+charts
+├── control-plane
+│   ├── charts
```

Comment:
I assume we’ll want helm to manage `charts` and do a `requirements.yaml` here, right?

Reply:
I thought `requirements.yaml` is only needed if the dependency (in this case, `charts/partials`) is published?

Reply:
Helm assumes that it manages the `charts` directory and it is only populated on `helm package`. We *could* hack it with some symlinks, and that might be the best solution, I just wanted to call out the one true way (tm).

You can have local paths in `requirements.yaml` as well, fwiw. No need to publish.

Reply:
Let me know if any of these makes sense:

When pushing the `control-plane` chart to the helm charts repo (https://github.com/helm/charts), some CI job there will package it to publish it to https://kubernetes-charts.storage.googleapis.com.
The only charts that job will have access to are the already published ones and the ones under `control-plane/charts`. For that reason it seems to me having a `requirements.yaml` file can't work here. OTOH when I test `helm package` with symlinks, it produces a tgz file that when unpacked will have `charts/partials` populated. I guess that unpacked directory is what we'd have to push to the helm charts repo.

Also note that `helm package` complains if the directory name doesn't match the chart name. So I'll have to rename `control-plane` to `linkerd2`.

Reply:
Makes sense to me to package up `partials` as .tgz file inside `control-plane/charts`, since we aren't publishing `partials`.

> So I'll have to rename control-plane to linkerd2.

Makes sense. The `data-plane` folder name may not even make sense; maybe just `patch`?

Reply:
@alpeb in the past I've written a `helm package` script that just copies the directory to `tmp` and renames it. I don't mind calling it `linkerd2`.

This raises a good point though, if we want this to be `helm install stable/linkerd2`, we won't be able to run `helm package` or push the compiled release anywhere. It will end up being in raw form in the stable repo (https://github.com/helm/charts/tree/master/stable/prometheus-adapter). This would require publishing both the partials *and* control-plane charts.

My proposal - don't worry about it, do what we need to do internally and write a publishing script for `stable/linkerd2` that copies the partials into the control-plane chart on "publish".

Reply:
I've just pushed a change with the charts renaming and a new task for the chart publication.
I've left the symlinks under `charts`. @grampelberg Let me know if you still deem necessary to have a `requirements.yml` file instead.